src/HttpRequest.cc

```diff
@@ -542,7 +542,8 @@ HttpRequest::maybeCacheable()
     // Because it failed verification, or someone bypassed the security tests
     // we cannot cache the response for sharing between clients.
     // TODO: update cache to store for particular clients only (going to same Host: and destination IP)
-    if (!flags.hostVerified && (flags.intercepted || flags.interceptTproxy))
+    // XXX: add missing checks, see HttpRequest::mustGoToOriginalDestination()
+    if (!flags.hostVerified && masterXaction->hasListeningInterceptedPort())
```

Comment:
> The current request may change independently from the port it was received on. eg. by adaptation/redirect making a new request message,

This code does not examine the _current_ port configuration (which may change at any time indeed). This code examines the port configuration that existed at master transaction creation time. Bugs notwithstanding, adapted and similarly adjusted requests inherit their master transaction object (and that original port configuration with it) from the original request. See HttpRequest::clone() and HttpRequest::inheritProperties().


This "inheritance" approach may be right or wrong, but the official code relies on the same logic: HttpRequest::flags are copied/inherited like the HttpRequest::masterXaction is. Any old bugs associated with this inheritance logic are probably out of scope here.

Even if we modify the code to compute the value of this `if` condition at the original HttpRequest object creation time, we will still have the same concern: A "derived" (adaptation, etc.) HttpRequest object may, in theory, forget to inherit that computed value. The only way to address that is to change the default from "assume cachable" to the opposite:

```C++
if (!flags.passedHostBasedCachabilityTest)
    return false; // either failed the test (OK) or forgot to copy the flag (bug)
```

However, that seems like an out of scope change to me.

Reply:
Neither of you are getting my point. Testing the properties of the port or the virgin HttpRequest are both wrong.

The flags on an HttpRequest coming out of adaptation may indicate that it was generated by the adaptator - making it cacheable. The flags to be checked should by the ones on the "this" HttpRequest object.

Reply:
> The flags on an HttpRequest coming out of adaptation may indicate that it was generated by the adaptator - making it cacheable.

If Eduard is [right](https://github.com/squid-cache/squid/pull/835#discussion_r682675764), then flags.intercepted and flags.interceptTproxy this code was examining before this PR were not affected by adaptation. Consequently, this PR preserves that official code logic. If that official "once intercepted, always intercepted" logic is wrong, we can fix it, but I hope we do not need to fix it in _this_ PR.

Thus, AFAICT, the key question here is whether adaptation changes flags.intercepted and flags.interceptTproxy.

Reply:
I have just checked the redirector code has become buggy with the flags clone function now being used. If adaptation is doing the same that is a bug that needs fixing. All I am requesting here is this PR to use the HttpRequest copy of flags and let such bugs be fixed at their source.

Reply:
> All I am requesting here is this PR to use the HttpRequest copy of flags

AFAICT, you are requesting to use RequestFlags members that do not exist in this PR (for a rather fundamental/good reason). Thus, you are effectively requesting to undo/redo a lot of serious, difficult, important PR changes[1]. Undoing these changes will have no material effects on the current Squid functionality. The only remaining motivation for this change request (that I can sense) is a future fix of an alleged bug that probably existed for a long time and that clearly requires serious consideration and work (to arrive at a _proper_ solution). That future bug fix may not even benefit from this change request at all (and might benefit from the PR changes)!

I can help with that future bug fix, but I think this change request is technically unfounded, humanly unfair and should be withdrawn. Please reconsider.

[1] Some of the relevant PR changes exposed problems in the existing code that we clearly did not know about and some of those problems were hidden by such bad design decisions as excessive use of poorly defined flags that the change request wants to restore. We should be celebrating these flags removal, even if it might make some future bug fixes a bit more difficult, not fighting tooth and nail to preserve them!